test/integration_test.py

```diff
@@ -404,17 +404,21 @@ def _get_url_content(self, url: str) -> bytes:
     def _get_url(self,
                  url: str,
                  allow_redirects: bool = True,
+                 follow_redirects: bool = True,
```

Comment:
This signature is too complicated. The distinction (and interaction) between the keyword arguments isn't clear to me.

What happens if you leave `_get_url` as is?

Reply:
When making an unauthorized request to the managed access file URL, `_get_url` fails with an `AssertionError` due to the unexpected 404 status. I suppose this could be worked around with `assertRaises`. When the request is authorized with the service account credentials, the behavior depends on whether `allow_redirects` is True or False. When it's True, `_get_url` _follows_ the redirects until it gets a 200 response and attempts to download the file (and in this case, fails with an `OverflowError` due to the large file size). When it's False, `_get_url` fails immediately with a `MaxRetryError`.

Reply:
On the previous commit, before changes were made to `_get_url`:
```
$ git log --oneline | head -1
39cc0381 Cover OAuth 2.0 authentication in integration test (#3266)
$ python3
Python 3.8.3 (default, May 22 2020, 15:24:35) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from integration_test import IndexingIntegrationTest; i = IndexingIntegrationTest(); i.setUp();
>>> file_url = 'https://service.dev.singlecell.gi.ucsc.edu/repository/files/2cc208d5-3025-4784-b15a-1c51bca94f0e?catalog=dcp2&version=2021-03-15T21%3A20%3A55.714000Z'
>>> i._get_url(file_url, allow_redirects=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 418, in _get_url
    self._assertResponseStatus(response, expected_statuses)
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 426, in _assertResponseStatus
    assert response.status in expected_statuses, (
AssertionError: (404, 'Not Found', b'{"Code":"NotFoundError","Message":"NotFoundError: Unable to find file \'2cc208d5-3025-4784-b15a-1c51bca94f0e\', version \'2021-03-15T21:20:55.714000Z\' in catalog \'dcp2\'"}')
>>> i._get_url(file_url, allow_redirects=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 418, in _get_url
    self._assertResponseStatus(response, expected_statuses)
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 426, in _assertResponseStatus
    assert response.status in expected_statuses, (
AssertionError: (404, 'Not Found', b'{"Code":"NotFoundError","Message":"NotFoundError: Unable to find file \'2cc208d5-3025-4784-b15a-1c51bca94f0e\', version \'2021-03-15T21:20:55.714000Z\' in catalog \'dcp2\'"}')
>>> with i._service_account_credentials():
...     i._get_url(file_url, allow_redirects=True)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 411, in _get_url
    response = self._http.request('GET',
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/request.py", line 75, in request
    return self.request_encode_url(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/request.py", line 97, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/auth/transport/urllib3.py", line 387, in urlopen
    response = self.http.urlopen(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/poolmanager.py", line 378, in urlopen
    return self.urlopen(method, redirect_location, **kw)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/poolmanager.py", line 336, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/connectionpool.py", line 690, in urlopen
    response = self.ResponseCls.from_httplib(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/response.py", line 601, in from_httplib
    resp = ResponseCls(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/response.py", line 259, in __init__
    self._body = self.read(decode_content=decode_content)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/response.py", line 515, in read
    data = self._fp.read() if not fp_closed else b""
  File "/home/nadove/.pyenv/versions/3.8.3/lib/python3.8/http/client.py", line 467, in read
    s = self._safe_read(self.length)
  File "/home/nadove/.pyenv/versions/3.8.3/lib/python3.8/http/client.py", line 608, in _safe_read
    data = self.fp.read(amt)
  File "/home/nadove/.pyenv/versions/3.8.3/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/contrib/pyopenssl.py", line 313, in recv_into
    return self.connection.recv_into(*args, **kwargs)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/OpenSSL/SSL.py", line 1733, in recv_into
    result = _lib.SSL_read(self._ssl, buf, nbytes)
OverflowError: integer 4347248640 does not fit '32-bit int'
>>> with i._service_account_credentials():
...     i._get_url(file_url, allow_redirects=False)
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/home/nadove/PycharmProjects/azul-sc/test/integration_test.py", line 411, in _get_url
    response = self._http.request('GET',
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/request.py", line 75, in request
    return self.request_encode_url(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/request.py", line 97, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/auth/transport/urllib3.py", line 387, in urlopen
    response = self.http.urlopen(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/poolmanager.py", line 365, in urlopen
    retries = retries.increment(method, url, response=response, _pool=conn)
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/urllib3/util/retry.py", line 439, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='service.dev.singlecell.gi.ucsc.edu', port=443): Max retries exceeded with url: https://service.dev.singlecell.gi.ucsc.edu/repository/files/2cc208d5-3025-4784-b15a-1c51bca94f0e?catalog=dcp2&version=2021-03-15T21%3A20%3A55.714000Z (Caused by ResponseError('too many redirects'))
```